lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+  tryConvertingReturnValue(const SortedVector<RegisterType> &UsedRegisters);
+};
+
+class Converter : public ConverterBase {
+public:
+  using ConverterBase::ConverterBase;
+  TypeDependencyManager manageTypeDependencies(model::Binary &Binary) const {
+    return TypeDependencyManager(ConverterBase::manageTypeDependencies(Binary));
+  }
+
+public:
+  /// Entry point for the `toCABI` conversion.
+  std::optional<model::TypePath> toCABI(const model::RawFunctionType &Function,
+                                        TupleTree<model::Binary> &Binary) const;
+};
+
+std::optional<model::TypePath>
+Converter::toCABI(const model::RawFunctionType &Function,
+                  TupleTree<model::Binary> &Binary) const {
+  revng_log(Log, "Converting a `RawFunctionType` to `CABIFunctionType`.");
+  revng_log(Log, "Function:\n" << serializeToString(Function));
+
+  if (!ABI.checkCompatibility(Function.Arguments(), Function.ReturnValues())) {
+    revng_log(Log,
+              "FAIL: the function is not compatible with `"
+                << model::ABI::getName(ABI.ABI()) << "`.");
+    return std::nullopt;
+  }
+
+  // A single place to conveniently store all the types that need to be added
+  // to the model in order for the new function prototype to be valid.
+  auto M = manageTypeDependencies(*Binary);
+
+  // Convert register arguments first.
+  auto RegArguments = M.tryConvertingRegisterArguments(Function.Arguments());
+  if (RegArguments == std::nullopt) {
+    revng_log(Log, "FAIL: Register argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // Then stack arguments.
+  auto Stack = M.tryConvertingStackArguments(Function.StackArgumentsType(),
+                                             RegArguments->size());
+  if (Stack == std::nullopt) {
+    revng_log(Log, "FAIL: Stack argument conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // And the return type.
+  auto ReturnType = M.tryConvertingReturnValue(Function.ReturnValues());
+  if (ReturnType == std::nullopt) {
+    revng_log(Log, "FAIL: Return value conversion is not possible.");
+    M.drop();
+    return std::nullopt;
+  }
+
+  // If all the partial conversions passed, the conversion is guaranteed to be
+  // successful. All the `TypeDependencies` can now be recorded into the
+  // model.
+  M.commit();
+
+  // With all the dependencies out of the way, new `CABIFunctionType` can now
+  // be created,
+  auto [NewType, NewTypePath] = Binary->makeType<model::CABIFunctionType>();
+  NewType.ID() = Function.ID();
+  NewType.CustomName() = Function.CustomName();
+  NewType.OriginalName() = Function.OriginalName();
+  NewType.ABI() = ABI.ABI();
+
+  // And filled in with the argument information.
+  using model::Argument;
+  for (auto &Argument : llvm::concat<Argument>(*RegArguments, *Stack))
+    NewType.Arguments().insert(std::move(Argument));
+  NewType.ReturnType() = *ReturnType;
+
+  revng_log(Log, "Conversion successful.");
```

Comment:
It was taking advantage of the fact that the only way for this message to be emitted is to go through the earlier log entries inside the same function that provide the said context. And if the flow of this linear-as-a-nail function somehow breaks - we have even bigger problems on our hands that broken longs.

But yeah, it can still benefit from extra indentation, I'll take a look.

Reply:
Updates?

Reply:
I sprinkled the indentation: https://github.com/revng/revng/pull/232/commits/ee2f59ecd0537a8bd18847276047573b16541e61

P. S. I can't express how badly I wanted something like
```cpp
struct ScopedIndent {
  std::size_t IndentSize;
  Logger &Log;
  ScopedIndent(std::size_t IndentSize) : IndentSize(IndentSize) { Log.indent(IndentSize); }
  ~ScopedIndent() { Log.unindent(IndentSize); }
};

class Logger {
// . . .
  ScopedIndent scopedIndent(std::size_t IndentSize = 1) { return ScopedIndent(IndentSize); }
```

Reply:
There's `LoggerIndent`, I mentioned it in my original comment

Reply:
Ah, whoops. I just read it as a straight-up english instead of noticing that it's actually a class name. My bad.

Reply:
It now uses `LoggerIndent` :tada: